[mta] Backward of unary foreach functions

crcrpar opened the pull request:
as per title, this PR defines backward of those.

This doesn't implement forward-mode automatic differentiation as [the current codegen](https://github.com/pytorch/pytorch/blob/a747326423ed4731996769e3b8eb73eecbdee2d4/tools/autograd/gen_variable_type.py#L1513) doesn't seem to handle `ArrayRef<Tensor>`.

Rel:
- https://github.com/pytorch/pytorch/issues/53796
- https://github.com/pytorch/pytorch/issues/58833

cc @mcarilli @ngimel

crcrpar commented:
regarding inplace ops, I locally confirmed the version seems to be bumped appropriately
```
In [8]: tensors = [torch.randn(3).requires_grad_() for _ in range(2)]
In [9]: print([t._version for t in tensors])
[0, 0]
In [10]: torch._foreach_exp_(tensors)
In [11]: print([t._version for t in tensors])
[1, 1]
```

ngimel commented:
So the backward of foreach will always fall back to slow version, and codegen would make sure that we properly error out if versions don't match? That looks fine, but I'll let @alband take a look.

albanD commented:
> regarding inplace ops, I locally confirmed the version seems to be bumped appropriately

I think this is only because you're on CPU: the slow code calls the single-Tensor op which properly bumps the version counter.
Running the same thing on CUDA won't work I think.

albanD commented on torch/_torch_docs.py:
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
+.. note::
+    Only CPU and CUDA are supported. Forward-mode AD is not supported.
+
+Args:
```
These args are not correct for functions that take more arguments (like round). Should we just skip it altogether here and just recommend to read the original doc where all Tensor input/output are replaced with list of Tensors?

crcrpar replied:
That sounds right but I'm not aware of any unary foreach function taking optional arguments, let me see

albanD commented on torch/_torch_docs.py:
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
+    Only CPU and CUDA are supported. Forward-mode AD is not supported.
+
+Args:
+    self (list of Tensors): Input list of Tensors. Each Tensor can have an arbitrary shape, dtype, device, and strides.
```
It works for difference dtype and device? Why do we collect grads to zero out by device/dtype here then: https://github.com/pytorch/pytorch/blob/274d3b24c3f4b6dc8964ee6642e909ce4b992112/torch/optim/optimizer.py#L394 ?

crcrpar replied:
To reduce the number of CUDA kernels launched as possible, I guess.

albanD replied:
I'm not sure to follow? How does the two have different number of kernel launched?

crcrpar replied:
If we pass a list of tensors of different dtypes and/or devices to a foreach function, it'll call the corresponding aten native function the number of Tensors times, i.e. len(self) cuda kernels. With the grouping, it could reduce the number to `len(per_device_and_dtype_grads)`, assuming all the tensors are a CUDA tensor

ngimel replied:
@alband _for_each ops support args on the different device and of the different dtypes, but they'll fall back to slow implementation. If, instead of just calling `for_each` we pre-sort the args, we can call `for_each` with a subset of tensors that'll go to the fast path. Really this sorting should be done in for_each itself though, and not in the optimizer

albanD replied:
Ok!
Should we really advertise this then as it will not do what you expect these functions to do.

albanD commented on torch/_torch_docs.py:
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
+    )
+
+add_docstr(
+    torch._foreach_zero_,
```
This one doesn't need to be special cased? torch.zero_ does exist.

crcrpar replied:
as `_foreach_zero` the outplace one doesn't exist.

albanD replied:
We can special case it above then to only generate these docs if the corresponding `torch._foreach_*` op exists?

crcrpar replied:
I don't understand. I handled torch._foreach_zero_ differently because there's no `torch._foreach_zero`. The for-loop above is to generate the docs of `torch._foreach_foo` and `torch._foreach_foo_`.

What would you recommend? Defining `torch._foreach_zero` for the parity with the other unary foreach functions?

albanD replied:
If the for loop above, you could have a guard for each definition: `if hasattr(torch, "_foreach_foo")` then add the doc for it.
That will remove the need to special case zero (and work for others as well in the future).

albanD reviewed (commented):
This is a lot of code duplication.
Should we just have the codegen take care of it? The same way we re-use the `foo` formula for `foo_` when no explicit formula for `foo_` is provided. We could codegen a slow `_foreach_foo` formula when it is not provided. WDYT?

crcrpar committed:
Define unary foreach functions' backward

with a new CodeTemplate of DERIVATIVE_SINGLE_FOREACH

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
docs

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

albanD commented on docs/source/torch.rst:
```diff
@@ -592,6 +592,73 @@ BLAS and LAPACK Operations
+~~~~~~~~~~~~~~~~~~
+
+.. warning::
+    This API is in beta and subject to future changes.
```
nit: beta or prototype?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2927,3 +2927,89 @@
+# TODO(crcrpar): Consider defining fast path with the help of `multi_tensor_apply`.
+# rel: `DERIVATIVE_SINGLE_FOREACH` CodeTemplate.
+- name: _foreach_exp(Tensor[] self) -> Tensor[]
+  self: grads[i] * result[i].conj()
```
This formula is a duplicate of the `exp()` function above.
I was thinking that we could extend https://github.com/pytorch/pytorch/blob/f7939b21e1f09521857515f913cd03d13ea2287b/torchgen/api/autograd.py#L325-L360 to re-use that existing formula and put that in the template you already have.

In particular you can redefine the `grad`, `result`, etc that is used in the original formula within the for loop in the template and just dump the formula in there.
WDYT?

albanD replied:
No you won't need to modify the native functions in any way.
This function is just updating the `formula` field everything else is already properly set.

You just need to generate a good `formula` string for it.

crcrpar replied:
I understand it's cleaner by handling things in codegen but
- We could implement a fast path for backward as well as forward.
- with that this PR will cover the whole foreach functions e.g. binary with `Scalar` and `Scalar[]`

albanD replied:
> We could implement a fast path for backward as well as forward, which would not be compatible with that way of defining backward

You can, this logic that generates the formula kicks in only if you didn't provide a formula already. That's why some inplace functions have formulas and others don't.

> with that this PR will cover the whole foreach functions e.g. binary with Scalar and Scalar[]

Not sure what you mean by that?

crcrpar committed:
clean up docs gen

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Codegen unary foreach functions' backward

Pushing the heavy lifting to torchgen

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
manually bump version of `TensorList self`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
Added a codegen and a version bump

albanD commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -7,6 +7,16 @@ namespace at { namespace native {
+// rel: https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482
+inline void maybe_increment_version(TensorList tensors) {
+  for (const auto & t : tensors) {
+    if (t.requires_grad() && at::GradMode::is_enabled()) {
```
This check is not ok. even things that don't require gradients might get saved for backward and thus we need to always bump their version counter.

albanD commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -73,6 +73,7 @@ template <typename scalar_t, template<class> class Op> void foreach_unary_op_(Te
                                          /* r_args_depth */ 1,
                                          /* res_arg_index */ 0>(),
                           Op<opmath_t>());
+    maybe_increment_version(tensors);
```
Do I misremember our discussion where you showed me an example where the version was bumped properly? Why does this need to be added?

crcrpar replied:
That happened when inputs tensors are a CPU tensor not when fast path is chosen.
Even the functions registered to native_functions.yaml with CUDA key could go into the slow path which just calls aten native function. Therefore I decided to put this manual version bump here

albanD replied:
Ho that's right!
I'm still not sure why the InplaceOrView kernel doesn't get generated automatically to do that already? But we can look into that later.

crcrpar replied:
I thought with https://github.com/pytorch/pytorch/blob/b33d9e2c87f9323e56672fe862fe1de704346a37/torchgen/model.py#L1388-L1405, inplace functions that modify (and return) multiple Tensors don't get caught by https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482

albanD reviewed (commented):
Thanks for the update!
All the codegen part looks good! Only questions about the InplaceOrView key side.

crcrpar committed:
bump version if `GradMode::is_enabled`

following https://github.com/pytorch/pytorch/pull/89591#discussion_r1073924159

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

albanD commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -7,6 +7,16 @@ namespace at { namespace native {
+// rel: https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482
+inline void maybe_increment_version(TensorList tensors) {
+  for (const auto & t : tensors) {
+    if (at::GradMode::is_enabled()) {
```
Ho sorry my comment wasn't clear enough.
The whole if was not ok. This is not file either. Inplace changes in no_grad mode must also be tracked.
In short, you must always bump the version, whatever the state of autograd.

crcrpar committed:
always bump version counter

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

albanD reviewed (approved):
Looks good!
Sorry for all the back and forth, this is definitely a challenging one...

crcrpar committed:
cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchbot merge

(knowing this won’t work this time but want to trigger more jobs)

albanD commented:
You have a `@pytorchbot label` if you need to add label (like the `ciflow/trunk` one) but you are not allowed.

crcrpar commented:
@pytorchbot merge
